### ISSUE TITLE
关于torch.optim.lr_scheduler.LinearLR参数total_iters的设置问题

### ISSUE THREAD

**Issue by lizhuoran1019** [agi-brain/xuance#135]
您好，我在理解代码中 `total_iters` 设置为 `self.config.running_steps` 这一点上存在一些困惑。

```python
class QMIX_Learner(LearnerMAS):
    def __init__(self,
                 config: Namespace,
                 model_keys: List[str],
                 agent_keys: List[str],
                 policy: nn.Module):
        ...
        self.scheduler = torch.optim.lr_scheduler.LinearLR(self.optimizer,
                                                           start_factor=1.0,
                                                           end_factor=self.end_factor_lr_decay,
                                                           total_iters=self.config.running_steps)
        ...
```

假设我共仿真100个episode，每个episode都有500个step，那么`self.config.running_steps` 应设置为 `500*100=50000`。
当设置`n_epochs=1，start_training=2500`并启用RNN时，从第5个episode开始，每采样完一个episode会执行一次`self.scheduler.step()`，那么scheduler的step最大为`100-5=95` 远小于 `self.cofnig.running_steps`。

我观察到这会导致**学习率几乎不下降**。
我不确定自己对于这部分代码的理解是否存在问题，希望您能解释一下为什么 total_iters 要设置为 self.config.running_steps ？
增加一个额外参数来设置`total_iters`是否更合适？

**Comment by wenzhangliu** [comment]
抱歉隔了这么久才回复。在您提出该Issue之前，我们也意识到了该问题，但是一直没有腾出手来解决。

这里的total_iters设置为self.config.running_steps确实不合适，因为前者表示update的次数，而后者是所有环境采样的总步数。

由于每次做实验的update次数不是固定的，特别是针对使用RNN的情况（与所有环境采样的总步数虽然固定，但是完整的回合数是不确定的），所以我们对total_iters做了一个估计，根据估计出来的数值对学习率进行衰减，相关改动见这两次提交（[``586cc03``](https://github.com/agi-brain/xuance/commit/586cc03674138e9b3394dcbea02db086f007e785), [``590b0f2``](https://github.com/agi-brain/xuance/commit/590b0f2897466461f787f52d7b877ac844c036cb)）。

最后，我们设置了self.end_factor_lr_decay默认为1.0，也即默认不设置学习率衰减。如果需要设置学习率衰减的话，只需要设置一下config文件里的``end_factor_lr_decay``即可，取值范围为``0-1.0``.
